src/hellogo_fares/read.rs

```diff
@@ -0,0 +1,988 @@
+            object_id: line_id,
+            perimeter_action: PerimeterAction::Included,
+        }
+    }
+}
+
+impl From<(String, (String, String))> for TicketUseRestriction {
+    fn from((ticket_use_id, (use_origin, use_destination)): (String, (String, String))) -> Self {
+        Self {
+            ticket_use_id,
+            restriction_type: RestrictionType::OriginDestination,
+            use_origin,
+            use_destination,
+        }
+    }
+}
+
+fn calculate_direct_price(distance_matrix_element: &Element) -> Result<Decimal> {
+    let distance_matrix_element_price = distance_matrix_element
+        .try_only_child("prices")?
+        .try_only_child("DistanceMatrixElementPrice")?;
+    Ok(utils::get_amount_units_factor(
+        distance_matrix_element_price,
+    )?)
+}
+
+fn get_distance(distance_matrix_element: &Element) -> Result<u32> {
+    let distance_str = distance_matrix_element.try_only_child("Distance")?.text();
+    distance_str
+        .parse()
+        .map_err(|_| format_err!("Failed to parse '{}' into a 'u32'", distance_str))
+}
+
+fn get_line_id(fare_frame: &Element, service_frame: &Element) -> Result<String> {
+    fn get_line_ref<'a>(fare_frame: &'a Element) -> Result<&'a str> {
+        let references: Vec<_> = fare_frame
+            .try_only_child("contentValidityConditions")?
+            .children()
+            .filter(|element| element.name() == "ValidityTrigger")
+            .flat_map(|validity_trigger| validity_trigger.children())
+            .filter(|element| element.name() == "TriggerObjectRef")
+            .filter(|trigger_object_ref| {
+                trigger_object_ref
+                    .try_attribute::<String>("nameOfRefClass")
+                    .map(|ref_class| ref_class == "Line")
+                    .unwrap_or(false)
+            })
+            .flat_map(|trigger_object_ref| trigger_object_ref.attr("ref"))
+            .collect();
+        if references.len() == 1 {
+            Ok(references[0])
+        } else {
+            bail!("Failed to find a Line reference")
+        }
+    }
+
+    fn get_line_id_from_line_ref(service_frame: &Element, line_ref: &str) -> Result<String> {
+        let values: Vec<String> = service_frame
+            .try_only_child("lines")?
+            .children()
+            .filter(|element| element.name() == "Line")
+            .filter(|line| {
+                line.try_attribute::<String>("id")
+                    .map(|id| id == line_ref)
+                    .unwrap_or(false)
+            })
+            .map(|line| utils::get_value_in_keylist(line, "KV1PlanningLijnNummer"))
+            .collect::<Result<_>>()?;
+        if values.len() == 1 {
+            Ok(values[0].clone())
+        } else {
+            bail!("Failed to find the Line with identifier '{}'", line_ref)
+        }
+    }
+
+    let line_ref = get_line_ref(fare_frame)?;
+    let line_id = get_line_id_from_line_ref(service_frame, line_ref)?;
```

Comment:
As a reminder, do not forget to check if the line exists in collections.lines.

Reply:
Thanks for the reminder. It's done in the new commit that introduce all the modification to handle prefixes (see 09154d71f7436db3c98c0ebe7fdc1c93342b55fd).